src/complex_fluids/ComplexFluidForcing.cpp

```diff
@@ -0,0 +1,957 @@
+                                        const int coarsest_ln,
+                                        const int finest_ln)
+{
+    for (int ln = coarsest_ln; ln <= finest_ln; ln++)
+    {
+        Pointer<PatchLevel<NDIM>> level = hierarchy->getPatchLevel(ln);
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            Pointer<Patch<NDIM>> patch = level->getPatch(p());
+            const Box<NDIM>& box = patch->getBox();
+            Pointer<CellData<NDIM, double>> data = patch->getPatchData(data_idx);
+            Pointer<CellData<NDIM, double>> log_data = patch->getPatchData(d_W_cc_idx);
+            for (CellIterator<NDIM> it(box); it; it++)
+            {
+                CellIndex<NDIM> i = *it;
+#if (NDIM == 2)
+                Eigen::Matrix2d mat;
+                mat(0, 0) = (*log_data)(i, 0);
+                mat(1, 0) = (*log_data)(i, 2);
+                mat(0, 1) = (*log_data)(i, 2);
+                mat(1, 1) = (*log_data)(i, 1);
+                mat = mat.exp();
+                (*data)(i, 0) = mat(0, 0);
+                (*data)(i, 1) = mat(1, 1);
+                (*data)(i, 2) = mat(0, 1);
+#endif
+#if (NDIM == 3)
+                Eigen::Matrix3d mat;
+                mat(0, 0) = (*log_data)(i, 0);
+                mat(1, 1) = (*log_data)(i, 1);
+                mat(2, 2) = (*log_data)(i, 2);
+                mat(1, 2) = mat(2, 1) = (*log_data)(i, 3);
+                mat(0, 2) = mat(2, 0) = (*log_data)(i, 4);
+                mat(0, 1) = mat(1, 0) = (*log_data)(i, 5);
+                mat = mat.exp();
+                (*data)(i, 0) = mat(0, 0);
+                (*data)(i, 1) = mat(1, 1);
+                (*data)(i, 2) = mat(2, 2);
+                (*data)(i, 3) = mat(1, 2);
+                (*data)(i, 4) = mat(0, 2);
+                (*data)(i, 5) = mat(0, 1);
+#endif
+            }
+        }
+    }
+    return;
+} // exponentiateMatrix
+
+void
+ComplexFluidForcing::projectMatrix(const int data_idx,
+                                   const Pointer<Variable<NDIM>> /*var*/,
+                                   const Pointer<PatchHierarchy<NDIM>> hierarchy,
+                                   const double /*data_time*/,
+                                   const bool initial_time,
+                                   const int coarsest_ln,
+                                   const int finest_ln)
+{
+    for (int ln = coarsest_ln; ln <= finest_ln; ln++)
+    {
+        Pointer<PatchLevel<NDIM>> level = hierarchy->getPatchLevel(ln);
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            Pointer<Patch<NDIM>> patch = level->getPatch(p());
+            const Box<NDIM>& box = patch->getBox();
+            if (initial_time) return;
+            Pointer<CellData<NDIM, double>> data = patch->getPatchData(data_idx);
+            for (CellIterator<NDIM> it(box); it; it++)
+            {
+                CellIndex<NDIM> i = *it;
+#if (NDIM == 2)
+                Eigen::Matrix2d c;
```

Comment:
Might be useful to have a helper function that will construct the tensor given the patch data and a cell index?